tests/testsuite/build_script_env.rs

```diff
@@ -5,6 +5,165 @@ use cargo_test_support::project;
+            "#,
+        )
+        .build();
+
+    p.cargo("check")
+        .with_stderr_data(str![[r#"
+[COMPILING] foo v0.1.0 ([ROOT]/foo)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    p.change_file(
+        ".cargo/config.toml",
+        r#"
+            [env]
+            FOO = "bad"
+        "#,
+    );
+
+    p.cargo("check")
+        .with_stderr_data(str![[r#"
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    p.cargo("clean").run();
+    p.cargo("check")
+        .with_status(101)
+        .with_stderr_data(
+            "\
+            [COMPILING] foo v0.1.0 ([ROOT]/foo)
+[ERROR] failed to run custom build command for `foo v0.1.0 ([..])`
+...",
+        )
+        .run();
+}
+
+#[cfg(windows)]
+#[cargo_test]
+fn rerun_if_env_changes_config_windows() {
+    let p = project()
+        .file("Cargo.toml", &basic_manifest("foo", "0.1.0"))
+        .file("src/main.rs", "fn main() {}")
+        .file(
+            ".cargo/config.toml",
+            r#"
+                [env]
+                FOO = "good"
+            "#,
+        )
+        .file(
+            "build.rs",
+            r#"
+                fn main() {
+                    println!("cargo:rerun-if-env-changed=Foo");
+                    if let Ok(foo) = std::env::var("Foo") {
+                        assert!(&foo != "bad");
+                    }
+                }
+            "#,
+        )
+        .build();
+
+    p.cargo("check")
+        .with_stderr_data(str![[r#"
+[COMPILING] foo v0.1.0 ([ROOT]/foo)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    p.change_file(
+        ".cargo/config.toml",
+        r#"
+            [env]
+            FOO = "bad"
+        "#,
+    );
+
+    p.cargo("check")
+        .with_stderr_data(str![[r#"
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    p.cargo("clean").run();
+    p.cargo("check")
+        .with_status(101)
+        .with_stderr_data(
+            "\
+[COMPILING] foo v0.1.0 ([ROOT]/foo)
+[ERROR] failed to run custom build command for `foo v0.1.0 ([..])`
+...",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn rerun_if_env_changes_cargo_set_env_variable() {
+    let p = project()
+        .file("src/main.rs", "fn main() {}")
+        .file(
+            "build.rs",
+            r#"
+                fn main() {
+                    println!("cargo:rerun-if-env-changed=OUT_DIR");
+                }
+            "#,
+        )
+        .build();
+
+    p.cargo("check")
+        .with_stderr_data(str![[r#"
+[COMPILING] foo v0.0.1 ([ROOT]/foo)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    p.change_file(
+        ".cargo/config.toml",
+        r#"
+            [env]
+            OUT_DIR = "somedir"
+        "#,
+    );
```

Comment:
The intention of this test is for detecting changes made by cargo, and not the user.  Us manually setting one runs counter to that

Reply:
I'm not too sure what the concern is.

Reply:
Setting `env.OUR_DIR` has no effect on build script execution.

Reply:
Oh, that.  The intent of my comment was to say that we should not test this with `[env]` but making a change that causes the cargo-provided env to change.